quesma/schema/hierarchical_schema.go

```diff
@@ -0,0 +1,47 @@
+// Copyright Quesma, licensed under the Elastic License 2.0.
+// SPDX-License-Identifier: Elastic-2.0
+package schema
+
+type SchemaTreeNode struct {
+	Name string // for example: "products" or "manufacturer" for "products.manufacturer", "" for root
+
+	// Either this is a leaf node (non-nil field) or a branch node (non-empty children)
+	Field    *Field
+	Children []*SchemaTreeNode
+}
+
+func addToTree(root *SchemaTreeNode, field *Field) {
+	components := field.PropertyName.Components()
+
+	leafComponent := components[len(components)-1]
+	components = components[:len(components)-1]
+
+	currentNode := root
+	for _, component := range components {
+		found := false
+		for _, child := range currentNode.Children {
+			if child.Name == component {
+				currentNode = child
+				found = true
+				break
+			}
+		}
+		if !found {
+			newNode := SchemaTreeNode{Name: component}
+			currentNode.Children = append(currentNode.Children, &newNode)
+			currentNode = &newNode
+		}
+	}
+
+	currentNode.Children = append(currentNode.Children, &SchemaTreeNode{Name: leafComponent, Field: field})
+}
+
+func SchemaToHierarchicalSchema(s *Schema) *SchemaTreeNode {
```

Comment:
I think it's worth putting a comment over this method about what it does (the "unflattening") with an example. 
PR description is clear, but here it'll be more accessible in the future :)

Reply:
Added comment in https://github.com/QuesmaOrg/quesma/pull/658/commits/5cc46998a0aecbb14e912433eb21d81593c8b1e8